src/rust/engine/fs/src/lib.rs

```diff
@@ -371,8 +371,30 @@ impl PosixFS {
     pool: Arc<ResettablePool>,
     ignore_patterns: Vec<String>,
   ) -> Result<PosixFS, String> {
-    let root: &Path = root.as_ref();
-    let canonical_root = root
+    let ignore = create_ignore(&ignore_patterns).map_err(|e| {
+      format!(
+        "Could not parse build ignore inputs {:?}: {:?}",
+        ignore_patterns, e
+      )
+    })?;
+    Ok(PosixFS {
+      root: Self::canonicalize(root)?,
+      pool: pool,
+      ignore: ignore,
+    })
+  }
+
+  pub fn clone_with_root<P: AsRef<Path>>(&self, root: P) -> Result<PosixFS, String> {
+    Ok(PosixFS {
+      root: Self::canonicalize(root)?,
+      pool: self.pool.clone(),
+      ignore: self.ignore.clone(),
```

Comment:
The `ignore` patterns make this a bit challenging I think... the patterns used here are relative to the root of the repository/buildroot, and so we construct the `Gitignore` instance with an empty root. But if you were going to be capturing from a subdirectory of the repository and you still wanted to apply the same patterns, you'd need to pass in the portion of the subdirectory relative to the buildroot (afaik).

This is to account for patterns like:
```
$ cat .gitignore
/please/ignore/this/long/specific/file.sh
```

I think that not applying ignore patterns to this type of capture would be reasonable, and would probably be fine for now. But if we're going to apply them, I think we'd need to re-construct them as relative to the buildroot (if the capture was under the buildroot), or ignore the patterns if the capture was outside the buildroot.

Reply:
This is still an issue I think.

Reply:
Oops! Done